File: ADA_MDC/1123.py
import numpy as np                            # import NumPy for numerical operations
import matplotlib.pyplot as plt               # import pyplot for plotting

# ------------------------------------------------------------
# Part 1: Ordinary Least Squares (unchanged)
# ------------------------------------------------------------

# Load the data from the text file into a NumPy array
data = np.loadtxt('3060309_MDC1.txt')         # read two-column data (x,y) from file
# Separate into x (first column) and y (second column)
x = data[:, 0]                                # extract x values
y = data[:, 1]                                # extract y values
# Determine the number of observations
n = len(x)                                    # count data points

# Compute the sample means of x and y
x_bar = np.mean(x)                            # mean of x
y_bar = np.mean(y)                            # mean of y
# Compute the sum of squared deviations of x
S_xx = np.sum((x - x_bar) ** 2)               # Σ(x - x̄)²
# Compute the sum of cross‑deviations between x and y
S_xy = np.sum((x - x_bar) * (y - y_bar))       # Σ(x - x̄)(y - ȳ)
# Calculate the OLS slope estimator b̂
b_hat = S_xy / S_xx                           # slope = S_xy / S_xx
# Calculate the OLS intercept estimator â
a_hat = y_bar - b_hat * x_bar                 # intercept = ȳ - b̂ x̄

# Compute residuals (difference between observed and fitted y)
res = y - (a_hat + b_hat * x)                 # residuals y - (a + bx)
# Sum of squared residuals
SSR = np.sum(res ** 2)                        # Σ residual²
# Estimate of the noise variance σ²
s2h = SSR / (n - 2)                    # SSR/(n-2)

# Variance of the slope estimator
var_b = s2h / S_xx                     # σ²/S_xx
# Variance of the intercept estimator
var_a = s2h * (1 / n + x_bar ** 2 / S_xx)  # σ²(1/n + x̄²/S_xx)
# Covariance between intercept and slope estimators
cov_ab = -x_bar * s2h / S_xx           # -x̄ σ²/S_xx
# Standard error of the intercept
se_a = np.sqrt(var_a)                         # √var_a
# Standard error of the slope
se_b = np.sqrt(var_b)                         # √var_b

# Print the key OLS results
print(f"Number of data points: {n}")           # display n
print(f"Estimated slope (b): {b_hat:.5f}")     # display b̂
print(f"Estimated intercept (a): {a_hat:.5f}") # display â
print(f"Estimated noise variance (σ²): {s2h:.5f}")  # display σ² estimate
print(f"Standard error of a: {se_a:.5f}")      # display SE of intercept
print(f"Standard error of b: {se_b:.5f}")      # display SE of slope
print(f"Covariance of (a, b): {cov_ab:.5f}")   # display covariance

# Create a new figure with specified dimensions
plt.figure(figsize=(10, 6))                    # new figure 10x6 inches
# Scatter plot of the data points in maroon, with size and transparency
plt.scatter(x, y, s=60, alpha=0.7, color='maroon', label='Data Points')  # plot points
# Prepare two x‑values for drawing the fitted line across the range
x_line = np.array([x.min(), x.max()])          # endpoints for line
# Compute corresponding y‑values on the fitted line
y_line = a_hat + b_hat * x_line                # y = a + b x_line
# Plot the fitted line as a dashed chrome‑yellow line
plt.plot(
    x_line, y_line, '--', linewidth=2, color='#FFA700',
    label=f'Fit: y = {a_hat:.3f} + {b_hat:.3f}x'
)  # plot fit line

# Label the x-axis
plt.xlabel('x', fontsize=14)                   # set x label
# Label the y-axis
plt.ylabel('y', fontsize=14)                   # set y label
# Add a title to the plot
plt.title('Ordinary Least Squares Fit', fontsize=16)  # set title
# Show the legend in the plot
plt.legend(fontsize=12)                        # display legend
# Add a subtle dashed grid with major and minor ticks
plt.grid(True, linestyle='--', linewidth=0.5, alpha=0.7)  # enable grid
plt.minorticks_on()                            # enable minor ticks
plt.tight_layout()                             # adjust spacing
plt.show()                                     # render the plot


def ol(arr):                                
    # return an array of ones with the same shape and data type as arr
    return np.ones(arr.shape, dtype=arr.dtype)  # mimic np.ones_like

# custom ui for arbitrary-dimensional arrays
def ui(flat_index, shape):
    # convert a flat index into a tuple of coordinate indices for an array of given shape
    idxs = []                                    # list to collect indices
    for dim in reversed(shape):                  # iterate dims last-first
        idxs.append(flat_index % dim)            # compute index in this dim
        flat_index //= dim                       # reduce flat_index
    return tuple(reversed(idxs))                 # reverse to normal order

# 1. Load data
data = np.loadtxt('3060309_MDC1.txt')           # read data again
x = data[:, 0]                                  # extract x
y = data[:, 1]                                  # extract y
sigma = ol(y)                            # constant errors array
N = len(x)                                      # count points

# 2. Manual least‑squares estimates for a0, b0
x_mean = np.mean(x)                             # mean of x
y_mean = np.mean(y)                             # mean of y
Sxx = np.sum((x - x_mean) ** 2)                 # Σ(x - x̄)²
Sxy = np.sum((x - x_mean) * (y - y_mean))       # Σ(x - x̄)(y - ȳ)
a0 = Sxy / Sxx                                  # slope a₀
b0 = y_mean - a0 * x_mean                       # intercept b₀

# residual variance & parameter uncertainties
resid = y - (a0 * x + b0)                       # residuals
chi2_ls = np.sum((resid / sigma) ** 2)          # χ² of LS
s2r = chi2_ls / (N - 2)                # variance estimate
sigma_a = np.sqrt(s2r / Sxx)           # SE of slope
sigma_b = np.sqrt(s2r * (1.0/N + x_mean**2 / Sxx))  
                                                # SE of intercept

# 3. Build (a,b) grid ±3σ around LS solution
n_a, n_b = 200, 200                             # grid sizes
a_vals = np.linspace(a0 - 3*sigma_a, a0 + 3*sigma_a, n_a)  # a grid
b_vals = np.linspace(b0 - 3*sigma_b, b0 + 3*sigma_b, n_b)  # b grid
A, B = np.meshgrid(a_vals, b_vals)              # meshgrid arrays

# 4. Compute χ² on the grid
# broadcast x and y across grid dims
res = y[:, None, None] - (A[None, :, :] * x[:, None, None] + B[None, :, :])  # residuals grid
Chi2 = np.sum((res / sigma[:, None, None]) ** 2, axis=0)  # sum over data pts

# 5. Find best‑fit and Δχ² using custom ui
flat_idx = np.argmin(Chi2)                      # index of min χ²
i_min, j_min = ui(flat_idx, Chi2.shape)  # row,col of min
a_best = A[i_min, j_min]                        # best-fit a
b_best = B[i_min, j_min]                        # best-fit b
chi2_min = Chi2[i_min, j_min]                   # minimal χ²
delta_chi2 = Chi2 - chi2_min                    # Δχ² field

# 6. Compute credible‑region bounds
levels = [2.30, 6.17, 11.8]                     # Δχ² thresholds
region_ranges = []                              # list of bounds
for lvl in levels:
    mask = delta_chi2 <= lvl                    # mask region
    a_min, a_max = A[mask].min(), A[mask].max()  # a bounds
    b_min, b_max = B[mask].min(), B[mask].max()  # b bounds
    region_ranges.append((a_min, a_max, b_min, b_max))  # store

# 7. Print results
print("Best‑fit parameters:")                    # header
print(f"  a = {a_best:.5f}")                     # print a_best
print(f"  b = {b_best:.5f}")                     # print b_best
print(f"  χ²_min = {chi2_min:.3f}\n")            # print χ²_min

for pct, (a_min, a_max, b_min, b_max) in zip([68.3, 95.4, 99.73], region_ranges):
    print(f"{pct:.1f}% credible region:")       # label
    print(f"  a ∈ [{a_min:.5f}, {a_max:.5f}]")    # a interval
    print(f"  b ∈ [{b_min:.5f}, {b_max:.5f}]\n")  # b interval

# 8. Plot Δχ² contours
plt.figure(figsize=(8, 6))                      # new figure
cs = plt.contour(                               # contour lines
    A, B, delta_chi2,
    levels=levels,
    colors=['maroon', 'orange', 'olive'],
    linewidths=[1.5, 1.5, 1.5]
)
plt.plot(a_best, b_best, 'k+', label='Best‑fit') # mark best-fit

plt.xlabel('a')                                 # x-label
plt.ylabel('b')                                 # y-label
plt.title(r'$\Delta\chi^2$ Contours for $(a,b)$')  # title
plt.clabel(cs, inline=True, fmt={2.30:'68.3%', 6.17:'95.4%', 11.8:'99.73%'})  # annotations

# 9. Increase axis range by 5% margin
x_min, x_max = a_vals.min(), a_vals.max()       # a limits
y_min, y_max = b_vals.min(), b_vals.max()       # b limits
x_margin = 0.05 * (x_max - x_min)                # margin a
y_margin = 0.05 * (y_max - y_min)                # margin b
plt.xlim(x_min - x_margin, x_max + x_margin)     # set xlim
plt.ylim(y_min - y_margin, y_max + y_margin)     # set ylim

plt.legend()                                    # show legend
plt.show()                                      # render contour plot

# ------------------------------------------------------------
# Part 2: Metropolis MCMC using those OLS parameters
# ------------------------------------------------------------

# Log‑likelihood function using σ² from the OLS stage
def log_likelihood(a, b):
    return -0.5 * np.sum((y - (a + b * x))**2) / s2h  # return log L

# Define uniform priors ±5 around a_hat and ±3 around b_hat
a_min, a_max = a_hat - 5.0, a_hat + 5.0          # a prior bounds
b_min, b_max = b_hat - 3.0, b_hat + 3.0          # b prior bounds

def log_prior(a, b):
    if a_min <= a <= a_max and b_min <= b <= b_max:
        return 0.0                                # flat prior inside
    return -np.inf                                # zero outside

def log_posterior(a, b):
    lp = log_prior(a, b)                         # compute lp
    return lp + log_likelihood(a, b) if np.isfinite(lp) else -np.inf

# MCMC hyperparameters
prop_a, prop_b = 0.05, 0.02                     # proposal std devs
steps, burn, thin = 40000, 10000, 10            # steps, burn-in, thinning

# Initialize chain at the OLS estimates
current_a, current_b = a_hat, b_hat             # start values
current_lp = log_posterior(current_a, current_b)  # initial log-posterior

chain = []                                      # to store samples
accept = 0                                      # acceptance counter
rng = np.random.default_rng(0)                  # seeded RNG

for _ in range(steps):                          # MCMC loop
    cand_a = rng.normal(current_a, prop_a)       # propose new a
    cand_b = rng.normal(current_b, prop_b)       # propose new b
    cand_lp = log_posterior(cand_a, cand_b)      # compute new log-posterior
    if rng.random() < np.exp(cand_lp - current_lp):
        current_a, current_b, current_lp = cand_a, cand_b, cand_lp  # accept
        accept += 1
    chain.append([current_a, current_b])         # record

chain = np.asarray(chain)                       # to array
posterior = chain[burn::thin]                   # apply burn-in and thinning

# Summary statistics of the posterior
mean_a, mean_b = posterior.mean(axis=0)         # posterior means
std_a, std_b = posterior.std(axis=0, ddof=1)    # posterior std devs

print("\nMCMC results:")                         # header
print("Acceptance rate:", accept / steps)        # print rate
print(f"Posterior mean a = {mean_a:.5f} ± {std_a:.5f}")  # print a ±
print(f"Posterior mean b = {mean_b:.5f} ± {std_b:.5f}")  # print b ±

# Corner‑style plots of the posterior
plt.figure()                                    # new fig for a
plt.hist(posterior[:, 0], bins=40, density=True,
         color='orange', edgecolor='black', linewidth=1.2)  # hist of a
plt.axvline(mean_a, linestyle='--', color='black')         # mean line
plt.title("Posterior of parameter a")          # title
plt.xlabel("a")                                # x-label
plt.ylabel("Probability density")              # y-label

plt.figure()                                    # new fig for b
plt.hist(posterior[:, 1], bins=40, density=True,
         color='orange', edgecolor='black', linewidth=1.2)  # hist of b
plt.axvline(mean_b, linestyle='--', color='black')         # mean line
plt.title("Posterior of parameter b")          # title
plt.xlabel("b")                                # x-label
plt.ylabel("Probability density")              # y-label

H, xedges, yedges = np.histogram2d(             # 2D hist of (a,b)
    posterior[:, 0], posterior[:, 1], bins=50, density=True
)
xcenters = 0.5 * (xedges[:-1] + xedges[1:])      # x bin centers
ycenters = 0.5 * (yedges[:-1] + yedges[1:])      # y bin centers
Hsort = np.sort(H.flatten())[::-1]               # sorted densities
Hcum = np.cumsum(Hsort) / Hsort.sum()            # cumulative
level68 = Hsort[np.searchsorted(Hcum, 0.683)]    # 68% level
level95 = Hsort[np.searchsorted(Hcum, 0.954)]    # 95% level

plt.figure()                                    # new fig for contour
plt.contourf(xcenters, ycenters, H.T,
             levels=[level95, level68, H.max()], alpha=0.6)  # filled contours
plt.scatter(posterior[:, 0], posterior[:, 1], s=4, alpha=0.4)  # scatter samples
plt.xlabel("a")                                # x-label
plt.ylabel("b")                                # y-label
plt.title("Joint posterior of a and b")        # title

plt.show()                                     # render final plot

File: ADA_MDC/1123.m
% Part 1: Ordinary Least Squares (unchanged)

% Load the data from the text file into an array
data = load('3060309_MDC1.txt');    % read two‑column data (x,y)
x = data(:,1);                      % extract x values
y = data(:,2);                      % extract y values
n = length(x);                      % count data points

% Compute the sample means of x and y
x_bar = mean(x);                    % mean of x
y_bar = mean(y);                    % mean of y

% Compute sums of squared and cross deviations
S_xx = sum((x - x_bar).^2);         % Σ(x - x̄)²
S_xy = sum((x - x_bar) .* (y - y_bar)); % Σ(x - x̄)(y - ȳ)

% Calculate the OLS slope and intercept estimators
b_hat = S_xy / S_xx;                % slope = S_xy / S_xx
a_hat = y_bar - b_hat * x_bar;      % intercept = ȳ - b̂ x̄

% Compute residuals and estimate noise variance
res = y - (a_hat + b_hat * x);      % residuals y - (a + b x)
SSR = sum(res.^2);                  % Σ residual²
sigma2_hat = SSR / (n - 2);         % SSR/(n-2)

% Variance and covariance of estimators
var_b = sigma2_hat / S_xx;                          % σ²/S_xx
var_a = sigma2_hat * (1/n + x_bar^2 / S_xx);        % σ²(1/n + x̄²/S_xx)
cov_ab = -x_bar * sigma2_hat / S_xx;                % -x̄ σ²/S_xx
se_a = sqrt(var_a);                                 % √var_a
se_b = sqrt(var_b);                                 % √var_b

% Print the key OLS results
fprintf('Number of data points: %d\n', n);
fprintf('Estimated slope (b): %.5f\n', b_hat);
fprintf('Estimated intercept (a): %.5f\n', a_hat);
fprintf('Estimated noise variance (σ²): %.5f\n', sigma2_hat);
fprintf('Standard error of a: %.5f\n', se_a);
fprintf('Standard error of b: %.5f\n', se_b);
fprintf('Covariance of (a, b): %.5f\n', cov_ab);

% Plot data and fitted line
figure('Position',[100 100 800 600]);
scatter(x, y, 60, 'MarkerFaceColor',[0.5 0 0], 'MarkerEdgeColor','none','MarkerFaceAlpha',0.7);
hold on;
x_line = [min(x), max(x)];                             % endpoints for line
y_line = a_hat + b_hat * x_line;                       % y = a + b x
plot(x_line, y_line, '--', 'LineWidth',2, 'Color',[1 0.647 0]);
hold off;
xlabel('x','FontSize',14);
ylabel('y','FontSize',14);
title('Ordinary Least Squares Fit','FontSize',16);
legend('Data Points', sprintf('Fit: y = %.3f + %.3fx', a_hat, b_hat),'FontSize',12);
grid on;
grid minor;


%% Part 2: Grid search and χ² contours

% helper for constant errors
sigma = ones_like(y);              % constant errors array
N = length(x);                     % data count

% 2. Manual least‑squares estimates for a0 (slope) and b0 (intercept)
x_mean = mean(x);                  
y_mean = mean(y);
Sxx = sum((x - x_mean).^2);
Sxy = sum((x - x_mean) .* (y - y_mean));
a0 = Sxy / Sxx;                    % slope estimate
b0 = y_mean - a0 * x_mean;         % intercept estimate

% residual variance & parameter uncertainties
resid = y - (a0 * x + b0);
chi2_ls = sum((resid ./ sigma).^2);
sigma2_resid = chi2_ls / (N - 2);
sigma_a = sqrt(sigma2_resid / Sxx);
sigma_b = sqrt(sigma2_resid * (1/N + x_mean^2 / Sxx));

% 3. Build (a,b) grid ±3σ around LS solution
n_a = 200; n_b = 200;
a_vals = linspace(a0 - 3*sigma_a, a0 + 3*sigma_a, n_a);
b_vals = linspace(b0 - 3*sigma_b, b0 + 3*sigma_b, n_b);
[A, B] = meshgrid(a_vals, b_vals);

% 4. Compute χ² on the grid
Chi2 = zeros(size(A));
for i = 1:n_b
    for j = 1:n_a
        res_grid = y - (A(i,j) * x + B(i,j));
        Chi2(i,j) = sum((res_grid ./ sigma).^2);
    end
end

% 5. Find best‑fit and Δχ² using custom unravel_index
[chi2_min, flat_idx] = min(Chi2(:));
idx = unravel_index(flat_idx, size(Chi2));   % [row, col]
i_min = idx(1); j_min = idx(2);
a_best = A(i_min, j_min);
b_best = B(i_min, j_min);
delta_chi2 = Chi2 - chi2_min;

% 6. Compute credible‑region bounds
levels = [2.30, 6.17, 11.8];
region_ranges = zeros(numel(levels),4);
for k = 1:numel(levels)
    mask = delta_chi2 <= levels(k);
    region_ranges(k,:) = [min(A(mask)), max(A(mask)), min(B(mask)), max(B(mask))];
end

% 7. Print results
fprintf('\nBest‑fit parameters:\n');
fprintf('  a = %.5f\n', a_best);
fprintf('  b = %.5f\n', b_best);
fprintf('  χ²_min = %.3f\n\n', chi2_min);
perc = [68.3, 95.4, 99.73];
for k = 1:numel(levels)
    fprintf('%.1f%% credible region:\n', perc(k));
    fprintf('  a ∈ [%.5f, %.5f]\n', region_ranges(k,1), region_ranges(k,2));
    fprintf('  b ∈ [%.5f, %.5f]\n\n', region_ranges(k,3), region_ranges(k,4));
end

% 8. Plot Δχ² contours
figure('Position',[200 200 800 600]);
[C,h] = contour(A, B, delta_chi2, levels, 'LineWidth',1.5);
hold on;
plot(a_best, b_best, 'k+', 'DisplayName','Best‑fit');
hold off;
xlabel('a');
ylabel('b');
title('\Delta\chi^2 Contours for (a,b)');
clabel(C,h,{'68.3%','95.4%','99.73%'});
legend;


%% Part 3: Metropolis MCMC using OLS parameters

% Log‑likelihood function using σ² from the OLS stage
log_likelihood = @(a,b) -0.5 * sum((y - (a + b*x)).^2) / sigma2_hat;

% Define uniform priors ±5 around a_hat and ±3 around b_hat
a_min = a_hat - 5.0;  a_max = a_hat + 5.0;
b_min = b_hat - 3.0;  b_max = b_hat + 3.0;

% Log‑prior and log‑posterior
log_prior = @(a,b) (a>=a_min && a<=a_max && b>=b_min && b<=b_max).*0 + ~(a>=a_min && a<=a_max && b>=b_min && b<=b_max).*-Inf;
log_posterior = @(a,b) log_prior(a,b) + log_likelihood(a,b);

% MCMC hyperparameters
prop_a = 0.05; prop_b = 0.02;
steps = 40000; burn = 10000; thin = 10;

% Initialize chain
current_a = a_hat; current_b = b_hat;
current_lp = log_posterior(current_a, current_b);
chain = zeros(steps,2);
accept = 0;
rng(0);

for t = 1:steps
    cand_a = current_a + prop_a * randn;
    cand_b = current_b + prop_b * randn;
    cand_lp = log_posterior(cand_a, cand_b);
    if rand < exp(cand_lp - current_lp)
        current_a = cand_a;
        current_b = cand_b;
        current_lp = cand_lp;
        accept = accept + 1;
    end
    chain(t,:) = [current_a, current_b];
end

posterior = chain(burn+1:thin:end, :);
mean_a = mean(posterior(:,1));  mean_b = mean(posterior(:,2));
std_a  = std(posterior(:,1), 1);  std_b  = std(posterior(:,2), 1);

% Print MCMC summary
fprintf('\nMCMC results:\n');
fprintf('Acceptance rate: %.4f\n', accept/steps);
fprintf('Posterior mean a = %.5f ± %.5f\n', mean_a, std_a);
fprintf('Posterior mean b = %.5f ± %.5f\n', mean_b, std_b);

% Corner‑style plots
figure;
histogram(posterior(:,1),40,'Normalization','pdf','EdgeColor','black','LineWidth',1.2);
hold on; xline(mean_a,'--k'); hold off;
title('Posterior of parameter a');
xlabel('a'); ylabel('Probability density');

figure;
histogram(posterior(:,2),40,'Normalization','pdf','EdgeColor','black','LineWidth',1.2);
hold on; xline(mean_b,'--k'); hold off;
title('Posterior of parameter b');
xlabel('b'); ylabel('Probability density');

% Joint posterior contour
[H, xedges, yedges] = histcounts2(posterior(:,1),posterior(:,2),50,'Normalization','pdf');
xcenters = (xedges(1:end-1)+xedges(2:end))/2;
ycenters = (yedges(1:end-1)+yedges(2:end))/2;
Hsort = sort(H(:),'descend');
Hcum = cumsum(Hsort)/sum(Hsort);
level68 = Hsort(find(Hcum>=0.683,1));
level95 = Hsort(find(Hcum>=0.954,1));

figure;
contourf(xcenters, ycenters, H', [level95, level68, max(H(:))]);
hold on;
scatter(posterior(:,1), posterior(:,2), 4, 'filled','MarkerFaceAlpha',0.4);
hold off;
xlabel('a'); ylabel('b');
title('Joint posterior of a and b');


%% Local helper functions

function out = ones_like(arr)
    out = ones(size(arr));
end

function idx = unravel_index(flat_index, shape)
    % Convert a linear index into subscripts for an array of size 'shape'
    rem = flat_index - 1;
    dims = numel(shape);
    idx = zeros(1,dims);
    for k = dims:-1:1
        idx(k) = mod(rem, shape(k));
        rem = floor(rem / shape(k));
    end
    idx = idx + 1;
end
